### PR TITLE
indexer: implement get_events API with EventQuery

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

Other than title, a couple of follow-ups:
- somehow BCS encoding for SuiMoveStruct does not work, I use serde_json for now and will change to BCS later
- CI tests for various event queries

## Test Plan 

Local run to make sure
- event population works
Screenshot 2023-03-10 at 6 31 39 PM

- basic event query results are expected, the timestamps are expected to be different b/c today SQLite uses [current FN local time](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority.rs#L1393) as the event time, while we should use the timestamp on checkpoint for consistency, which is what indexer uses today.

```
curl --location --request POST http://127.0.0.1:9000 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getEvents", "params":[ {"EventType": "MoveEvent"} , null, 2, null]}'
{"jsonrpc":"2.0","result":{"data":[{"timestamp":1678485812944,"txDigest":"BqWMouXCepQLSFk2uXJpcSgtk7NXVhtYTb4g4FcKSRpk","id":{"txDigest":"BqWMouXCepQLSFk2uXJpcSgtk7NXVhtYTb4g4FcKSRpk","eventSeq":2},"event":{"type":"moveEvent","content":{"packageId":"0x0000000000000000000000000000000000000000000000000000000000000002","transactionModule":"devnet_nft","sender":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","type":"0x2::devnet_nft::MintNFTEvent","fields":{"creator":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","name":"Example NFT","object_id":"0xa749c5f5cea49dad7473d78722b10e68fab3982edf6b32be17369ecaee41a26a"},"bcs":"p0nF9c6kna10c9eHIrEOaPqzmC7fazK+Fzaeyu5Bomr5DCjN4gDI6cd5d2SB1v4O0450kXclStsLRKbCDpiA9wtFeGFtcGxlIE5GVA=="}}},{"timestamp":1678485813958,"txDigest":"CqjGquzm4jrK7GPedg7eo6QvyeCma9Xbu97DHJAYijQu","id":{"txDigest":"CqjGquzm4jrK7GPedg7eo6QvyeCma9Xbu97DHJAYijQu","eventSeq":2},"event":{"type":"moveEvent","content":{"packageId":"0x0000000000000000000000000000000000000000000000000000000000000002","transactionModule":"devnet_nft","sender":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","type":"0x2::devnet_nft::MintNFTEvent","fields":{"creator":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","name":"Example NFT","object_id":"0xd4c2c89e603e832e0bec45c39b1e0fed416be76925fd22267aaaf5c9443917af"},"bcs":"1MLInmA+gy4L7EXDmx4P7UFr52kl/SImeqr1yUQ5F6/5DCjN4gDI6cd5d2SB1v4O0450kXclStsLRKbCDpiA9wtFeGFtcGxlIE5GVA=="}}}],"nextCursor":{"txDigest":"CqjGquzm4jrK7GPedg7eo6QvyeCma9Xbu97DHJAYijQu","eventSeq":2},"hasNextPage":true},"id":1}%                                                   
curl --location --request POST http://127.0.0.1:3030 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "id":1, "method":"sui_getEvents", "params":[ {"EventType": "MoveEvent"} , null, 2, null]}'
{"jsonrpc":"2.0","result":{"data":[{"timestamp":1678485813663,"txDigest":"BqWMouXCepQLSFk2uXJpcSgtk7NXVhtYTb4g4FcKSRpk","id":{"txDigest":"BqWMouXCepQLSFk2uXJpcSgtk7NXVhtYTb4g4FcKSRpk","eventSeq":2},"event":{"type":"moveEvent","content":{"packageId":"0x0000000000000000000000000000000000000000000000000000000000000002","transactionModule":"devnet_nft","sender":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","type":"0x2::devnet_nft::MintNFTEvent","fields":{"creator":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","name":"Example NFT","object_id":"0xa749c5f5cea49dad7473d78722b10e68fab3982edf6b32be17369ecaee41a26a"},"bcs":"p0nF9c6kna10c9eHIrEOaPqzmC7fazK+Fzaeyu5Bomr5DCjN4gDI6cd5d2SB1v4O0450kXclStsLRKbCDpiA9wtFeGFtcGxlIE5GVA=="}}},{"timestamp":1678485814657,"txDigest":"CqjGquzm4jrK7GPedg7eo6QvyeCma9Xbu97DHJAYijQu","id":{"txDigest":"CqjGquzm4jrK7GPedg7eo6QvyeCma9Xbu97DHJAYijQu","eventSeq":2},"event":{"type":"moveEvent","content":{"packageId":"0x0000000000000000000000000000000000000000000000000000000000000002","transactionModule":"devnet_nft","sender":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","type":"0x2::devnet_nft::MintNFTEvent","fields":{"creator":"0xf90c28cde200c8e9c779776481d6fe0ed38e749177254adb0b44a6c20e9880f7","name":"Example NFT","object_id":"0xd4c2c89e603e832e0bec45c39b1e0fed416be76925fd22267aaaf5c9443917af"},"bcs":"1MLInmA+gy4L7EXDmx4P7UFr52kl/SImeqr1yUQ5F6/5DCjN4gDI6cd5d2SB1v4O0450kXclStsLRKbCDpiA9wtFeGFtcGxlIE5GVA=="}}}],"nextCursor":{"txDigest":"CqjGquzm4jrK7GPedg7eo6QvyeCma9Xbu97DHJAYijQu","eventSeq":2},"hasNextPage":true},"id":1}%   
```

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-indexer/migrations/2022-11-18-195259_transactions/up.sql [hunk 1]
```diff
@@ -1,12 +1,16 @@
-CREATE TABLE transactions
-(
+-- SuiAddress and ObjectId type, 0x + 64 chars hex string
+CREATE DOMAIN address VARCHAR(66);
+-- Max char length for base58 encoded digest
+CREATE DOMAIN base58digest VARCHAR(44);
```
Comment (gegaowp) [inline crates/sui-indexer/migrations/2022-11-18-195259_transactions/up.sql:4]: move the type definition here b/c migrations are run sequentially, otherwise events cannot use this type.

File: crates/sui-indexer/src/models/transactions.rs [hunk 2]
```diff
@@ -26,7 +26,7 @@ pub struct Transaction {
     pub recipients: Vec<Option<String>>,
     pub checkpoint_sequence_number: i64,
     pub transaction_time: Option<NaiveDateTime>,
-    pub transaction_kind: Option<String>,
+    pub transaction_kinds: Vec<Option<String>>,
```
Comment (666lcz) [inline crates/sui-indexer/src/models/transactions.rs:29]: why is this changed?

File: crates/sui-indexer/src/schema.rs [hunk 3]
```diff
@@ -194,7 +199,7 @@ diesel::table! {
         recipients -> Array<Nullable<Text>>,
         checkpoint_sequence_number -> Int8,
         transaction_time -> Nullable<Timestamp>,
-        transaction_kind -> Nullable<Text>,
+        transaction_kinds -> Array<Nullable<Text>>,
```
Comment (666lcz) [inline crates/sui-indexer/src/schema.rs:202]: ditto

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 4]
```diff
@@ -108,6 +114,119 @@ impl IndexerStore for PgIndexerStore {
...
+            _ => {
+                return Err(IndexerError::NotImplementedError(
+                    "Querying events by Recipient and Object is deprecated.".to_string(),
+                ));
+            }
+        }
+        // fetch one more item to tell if there is next page
+        let mut page_limit = limit
+            .unwrap_or(MAX_EVENT_PAGE_SIZE)
+            .min(MAX_EVENT_PAGE_SIZE)
```
Comment (666lcz) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:180]: nit: I would be more explicit here by returning error if limit exceeds `MAX_EVENT_PAGE_SIZE`

**Review by 666lcz — COMMENTED** [review]
Overall looks good to me! Just a few questions

File: crates/sui-indexer/migrations/2022-11-18-220045_events/up.sql [hunk 5]
```diff
@@ -1,11 +1,25 @@
...
+    event_bcs BYTEA NOT NULL
 );
 
 CREATE INDEX events_transaction_digest ON events (transaction_digest);
+CREATE INDEX events_sender ON events (sender);
+CREATE INDEX events_package ON events (package);
+CREATE INDEX events_module ON events (module);
+CREATE INDEX events_event_type ON events (event_type);
 CREATE INDEX events_event_time ON events (event_time);
+CREATE INDEX events_event_time_ms ON events (event_time_ms);
```
[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/migrations/2022-11-18-220045_events/up.sql:25]: on wave 2, I found for ppl analytics, for example for tokenmomics, they preferred `event_time` b/c its more intuitive, for JSON RPC queries, having `event_time_ms` would save us converting to `NaiveDateTime` each time.

File: crates/sui-indexer/src/schema.rs (hunk 3 above)
Comment (gegaowp) [reply crates/sui-indexer/src/schema.rs]: there was a schema mismatch, Patrick fixed it in a different direction.
Still making the change here from Nullable<Text> to Text b/c Text is the generated type while running `diesel database reset`, had to change a couple of other places to make it build.

File: crates/sui-indexer/migrations/2022-11-18-220045_events/up.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (666lcz) [reply crates/sui-indexer/migrations/2022-11-18-220045_events/up.sql:25]: I don't think duplicating data in the same table is a good practice and it should be cheap to convert to NaiveDateTime. This is also confusing and makes people wonder what's the difference, can we remove one of them?

**Review by 666lcz — APPROVED** [review]
Thanks for addressing the comments!

File: crates/sui-indexer/migrations/2022-11-18-220045_events/up.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (gegaowp) [reply crates/sui-indexer/migrations/2022-11-18-220045_events/up.sql]: sure, removed the NaiveDateTime then.
